File: server/channels/api4/ai_formatter.go
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

package api4

import (
	"encoding/json"
	"net/http"

	"github.com/mattermost/mattermost/server/public/shared/mlog"
	"github.com/mattermost/mattermost/server/v8/channels/app"
	"github.com/mattermost/mattermost/server/v8/channels/app/openai"
)

// FormatPreviewRequest represents the API request body for formatting preview
type FormatPreviewRequest struct {
	Message            string `json:"message"`
	Profile            string `json:"profile,omitempty"`
	CustomInstructions string `json:"custom_instructions,omitempty"`
}

// FormatApplyRequest represents the API request body for applying formatting
type FormatApplyRequest struct {
	Message            string `json:"message"`
	Profile            string `json:"profile,omitempty"`
	CustomInstructions string `json:"custom_instructions,omitempty"`
}

func (api *API) initFormatterRoutes() {
	api.BaseRoutes.AI.Handle("/format/preview", api.APISessionRequired(formatPreview)).Methods("POST")
	api.BaseRoutes.AI.Handle("/format/apply", api.APISessionRequired(formatApply)).Methods("POST")
	api.BaseRoutes.AI.Handle("/format/profiles", api.APISessionRequired(getFormattingProfiles)).Methods("GET")
}

// formatPreview handles POST /api/v4/ai/format/preview
func formatPreview(c *Context, w http.ResponseWriter, r *http.Request) {
	if !requireAIEnabled(c) {
		return
	}

	var req FormatPreviewRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		c.SetInvalidParamWithErr("body", err)
		return
	}

	// Validate request
	if req.Message == "" {
		c.SetInvalidParam("message")
		return
	}

	// Convert profile string to FormattingProfile
	profile := openai.FormattingProfessional
	if req.Profile != "" {
		if !app.IsValidFormattingProfile(req.Profile) {
			c.SetInvalidParam("profile")
			return
		}
		profile = openai.FormattingProfile(req.Profile)
	} else {
		// Get user's default profile from preferences
		preferences, err := c.App.GetOrCreateAIPreferences(c.AppContext.Session().UserId)
		if err == nil && preferences.FormattingProfile != "" {
			profile = openai.FormattingProfile(preferences.FormattingProfile)
		}
	}

	formattingReq := &app.FormattingRequest{
		Message:            req.Message,
		Profile:            profile,
		CustomInstructions: req.CustomInstructions,
	}

	response, err := c.App.PreviewFormatting(c.AppContext, formattingReq)
	if err != nil {
		c.Err = err
		return
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(response); err != nil {
		c.Logger.Error("Failed to encode format preview response", mlog.Err(err))
	}
}

// formatApply handles POST /api/v4/ai/format/apply
func formatApply(c *Context, w http.ResponseWriter, r *http.Request) {
	if !requireAIEnabled(c) {
		return
	}

	var req FormatApplyRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		c.SetInvalidParamWithErr("body", err)
		return
	}

	// Validate request
	if req.Message == "" {
		c.SetInvalidParam("message")
		return
	}

	// Convert profile string to FormattingProfile
	profile := openai.FormattingProfessional
	if req.Profile != "" {
		if !app.IsValidFormattingProfile(req.Profile) {
			c.SetInvalidParam("profile")
			return
		}
		profile = openai.FormattingProfile(req.Profile)
	} else {
		// Get user's default profile from preferences
		preferences, err := c.App.GetOrCreateAIPreferences(c.AppContext.Session().UserId)
		if err == nil && preferences.FormattingProfile != "" {
			profile = openai.FormattingProfile(preferences.FormattingProfile)
		}
	}

	formattingReq := &app.FormattingRequest{
		Message:            req.Message,
		Profile:            profile,
		CustomInstructions: req.CustomInstructions,
	}

	response, err := c.App.FormatMessage(c.AppContext, formattingReq)
	if err != nil {
		c.Err = err
		return
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(response); err != nil {
		c.Logger.Error("Failed to encode format apply response", mlog.Err(err))
	}
}

// getFormattingProfiles handles GET /api/v4/ai/format/profiles
func getFormattingProfiles(c *Context, w http.ResponseWriter, r *http.Request) {
	if !requireAIEnabled(c) {
		return
	}

	profiles, err := c.App.GetFormattingProfiles(c.AppContext)
	if err != nil {
		c.Err = err
		return
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(profiles); err != nil {
		c.Logger.Error("Failed to encode formatting profiles response", mlog.Err(err))
	}
}

File: webapp/channels/src/actions/ai_formatter.ts
// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {DispatchFunc, GetStateFunc} from '@mattermost/types/actions';
import type {FormatMessageRequest, FormatMessageResponse, FormattingProfileInfo} from 'types/ai';

import {Client4} from '@mattermost/client';

import {aiClient} from 'client/ai';

import AIActionTypes from 'utils/constants/ai';
import {logError} from 'mattermost-redux/actions/errors';

export function formatPreview(message: string, profile?: string, customInstructions?: string) {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        dispatch({type: AIActionTypes.AI_FORMAT_PREVIEW_REQUEST});

        try {
            const request: FormatMessageRequest = {
                message,
                profile,
                custom_instructions: customInstructions,
            };

            const response = await aiClient.formatPreview(request);
            dispatch({
                type: AIActionTypes.AI_FORMAT_PREVIEW_SUCCESS,
                data: response,
            });

            return {data: response};
        } catch (error) {
            dispatch({
                type: AIActionTypes.AI_FORMAT_PREVIEW_FAILURE,
                error: error instanceof Error ? error.message : 'Failed to preview formatting',
            });
            dispatch(logError(error as Error));
            return {error};
        }
    };
}

export function formatApply(message: string, profile?: string, customInstructions?: string) {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        dispatch({type: AIActionTypes.AI_FORMAT_APPLY_REQUEST});

        try {
            const request: FormatMessageRequest = {
                message,
                profile,
                custom_instructions: customInstructions,
            };

            const response = await aiClient.formatApply(request);
            dispatch({
                type: AIActionTypes.AI_FORMAT_APPLY_SUCCESS,
                data: response,
            });

            return {data: response};
        } catch (error) {
            dispatch({
                type: AIActionTypes.AI_FORMAT_APPLY_FAILURE,
                error: error instanceof Error ? error.message : 'Failed to apply formatting',
            });
            dispatch(logError(error as Error));
            return {error};
        }
    };
}

export function getFormattingProfiles() {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        dispatch({type: AIActionTypes.AI_GET_FORMATTING_PROFILES_REQUEST});

        try {
            const profiles = await aiClient.getFormattingProfiles();
            dispatch({
                type: AIActionTypes.AI_GET_FORMATTING_PROFILES_SUCCESS,
                data: profiles,
            });

            return {data: profiles};
        } catch (error) {
            dispatch({
                type: AIActionTypes.AI_GET_FORMATTING_PROFILES_FAILURE,
                error: error instanceof Error ? error.message : 'Failed to load formatting profiles',
            });
            dispatch(logError(error as Error));
            return {error};
        }
    };
}

export function clearFormatPreview() {
    return {
        type: AIActionTypes.AI_CLEAR_FORMAT_PREVIEW,
    };
}

